examples/common/tracing/TraceHandlers.cpp

```diff
@@ -73,42 +86,215 @@ class TraceOutput
     TraceStream * mStream = nullptr;
 };
 
-TraceOutput output;
+struct TraceHandlerContext
+{
+    TraceOutput * sink;
+};
 
-// TODO: Framework this into a registry of handlers for different message types.
-bool TraceDefaultHandler(const TraceEventFields & trace)
+TraceOutput gTraceOutput;
+TraceHandlerContext gTraceHandlerContext = { .sink = &gTraceOutput };
+
+std::string AsJsonKey(const std::string & key, const std::string & value, bool prepend_comma)
+{
+    return (prepend_comma ? ", "s : ""s) + "\""s + key + "\": " + value;
+}
+
+std::string AsFirstJsonKey(const std::string & key, const std::string & value)
+{
+    return AsJsonKey(key, value, /* prepend_comma = */false);
+}
+
+std::string AsNextJsonKey(const std::string & key, const std::string & value)
+{
+    return AsJsonKey(key, value, /* prepend_comma = */true);
+}
+
+std::string AsJsonString(const std::string & value)
+{
+    return "\""s + value + "\""s;
+}
+
+std::string AsJsonString(const Transport::PeerAddress * peerAddress)
+{
+    char convBuf[Transport::PeerAddress::kMaxToStringSize] = {0};
+
+    peerAddress->ToString(convBuf);
+    return convBuf;
+}
+
+std::string AsJsonBool(bool isTrue)
+{
+    return isTrue ? "true"s : "false"s;
+}
+
+std::string AsJsonHexString(const uint8_t * buf, size_t bufLen)
+{
+    // Fill a string long enough for the hex conversion, that will be overwritten
+    std::string hexBuf(2 * bufLen, '0');
+
+    Encoding::BytesToLowercaseHexBuffer(buf, bufLen, hexBuf.data(), hexBuf.size());
+
+    return AsJsonString(hexBuf);
+}
+
+std::string PacketHeaderToJson(const PacketHeader * packetHeader)
+{
+    std::string jsonBody;
+
+    uint32_t messageCounter = packetHeader->GetMessageCounter();
+    const Optional<NodeId> & sourceNodeId = packetHeader->GetSourceNodeId();
+    const Optional<NodeId> & destNodeId = packetHeader->GetDestinationNodeId();
+    const Optional<GroupId> & groupId = packetHeader->GetDestinationGroupId();
+    uint16_t sessionId = packetHeader->GetSessionId();
+    uint8_t messageFlags = packetHeader->GetMessageFlags();
+    uint8_t securityFlags = packetHeader->GetSecurityFlags();
+
+    jsonBody += AsFirstJsonKey("msg_counter", std::to_string(messageCounter));
+
+    if (sourceNodeId.HasValue())
+    {
+        jsonBody += AsNextJsonKey("source_node_id", std::to_string(sourceNodeId.Value()));
+    }
+
+    if (packetHeader->IsValidGroupMsg() && groupId.HasValue())
```

Comment:
is there a possibility of 'isValidGroupMsg' without a group id? Should we add some json data in that case?

Reply:
I can do that